### PR TITLE
delayed expiry of status keys in Redis

### PR DESCRIPTION
**Pull request by skaes**
By actually tracking handler execution in one of our applications,
we have proof that every once in a while business logic (i.e. the
process method of a message handler) gets executed more than once.

This patch is an experiment to see whether we can eliminate these
duplicates by shifting the responsibility of deleting the corresponding
status key to Redis itself (with a configurable grace period).

One possible source of duplicate handler executions is the application
crashing while the ack message for the successfully processed RabbitMQ
message is still sitting in a TCP send buffer. This complication would
be handled by the patch by tuning the grace period (in many cases).
However, to fix it in all cases we'd need to keep the status keys which
act as tombstones, for as long as a message hasn't officially expired.

### CODE REVIEW

File: test/beetle/message_test.rb [hunk 1]
```diff
@@ -127,33 +128,42 @@ def setup
...
+      status_key = keys.shift
+      assert @store.redis.exists(status_key)
+      assert @store.redis.ttl(status_key) <= @config.redis_status_key_expiry_interval
+      keys.each do |key|
         assert !@store.redis.exists(key)
       end
     end
 
-    test "succesful processing of a redundant message twice should delete all keys from the database" do
+    test "succesful processing of a redundant message twice should delete all keys from the database (except the staus key, which should be set to expire)" do
```
Comment (boosty) [inline test/beetle/message_test.rb]: `s/succesful/successful`

Comment (boosty) [reply test/beetle/message_test.rb]: `s/staus/status`

File: test/beetle/message_test.rb [hunk 2]
```diff
@@ -127,33 +128,42 @@ def setup
...
-    test "successful processing of a non redundant message should delete all keys from the database" do
+    test "successful processing of a non redundant message should delete all keys from the database (except the staus key, which should be set to expire)" do
       header = header_with_params({})
       header.expects(:ack)
       message = Message.new("somequeue", header, 'foo', :store => @store)
+      message.stubs(:simple?).returns(false)
 
       assert !message.expired?
       assert !message.redundant?
+      assert !message.simple?
```
Comment (boosty) [inline test/beetle/message_test.rb]: This assertion is just testing the stub above, which is unnecessary.

File: test/beetle/message_test.rb [hunk 3]
```diff
@@ -127,33 +128,42 @@ def setup
...
       assert message.redundant?
+      assert !message.simple?
 
       message.process(lambda {|*args|})
       message.process(lambda {|*args|})
 
-      @store.keys(message.msg_id).each do |key|
+      keys = @store.keys(message.msg_id)
+      status_key = keys.shift
+      assert @store.redis.exists(status_key)
```
Comment (boosty) [inline test/beetle/message_test.rb]: For completeness and clarity, we could add the assertion for the expected expiry/ttl here too (as in the test above).

File: lib/beetle/configuration.rb [hunk 4]
```diff
@@ -35,6 +35,12 @@ class Configuration
     # handler timeout.
     attr_accessor :redis_failover_timeout
 
+    # how long we want status keys survive after we are we have seen the second message of
+    # a redundant message pair. Setting this to a high value (hours) will reduce the
+    # likelyhood of executing handler logic, but require a lerger redis instance and cause
```
Comment (boosty) [inline lib/beetle/configuration.rb]: `s/likelyhood/likelihood`

Comment (boosty) [reply lib/beetle/configuration.rb]: `s/lerger/larger`

Comment (boosty) [reply lib/beetle/configuration.rb]: What does "larger redis instance" mean? I think we only mean larger in memory, so we should be clear.

File: lib/beetle/configuration.rb [hunk 5]
```diff
@@ -35,6 +35,12 @@ class Configuration
     # handler timeout.
     attr_accessor :redis_failover_timeout
 
+    # how long we want status keys survive after we are we have seen the second message of
```
[the remark this answers is not in the document]
Comment (skaes) [reply lib/beetle/configuration.rb]: fixed

[the remark this answers is not in the document]
Comment (skaes) [reply lib/beetle/configuration.rb]: fixed

File: lib/beetle/configuration.rb (hunk 4 above)
Comment (skaes) [reply lib/beetle/configuration.rb]: fixed

File: test/beetle/message_test.rb (hunk 1 above)
Comment (skaes) [reply test/beetle/message_test.rb]: fixed

File: test/beetle/message_test.rb (hunk 2 above)
Comment (skaes) [reply test/beetle/message_test.rb]: fixed

File: test/beetle/message_test.rb (hunk 3 above)
Comment (skaes) [reply test/beetle/message_test.rb]: done

File: test/beetle/message_test.rb [hunk 6]
```diff
@@ -127,33 +128,42 @@ def setup
...
+      assert @store.redis.ttl(status_key) <= @config.redis_status_key_expiry_interval
+      keys.each do |key|
         assert !@store.redis.exists(key)
       end
     end
 
-    test "succesful processing of a redundant message twice should delete all keys from the database" do
+    test "successful processing of a redundant message twice should delete all keys from the database (except the staus key, which should be set to expire)" do
```
Comment (boosty) [inline test/beetle/message_test.rb]: Still says `staus` 😉

File: lib/beetle/configuration.rb [hunk 7]
```diff
@@ -35,6 +35,12 @@ class Configuration
     # handler timeout.
     attr_accessor :redis_failover_timeout
 
+    # how long we want status keys to survive after we have seen the second message of
+    # a redundant message pair. Setting this to a high value (hours) will reduce the
+    # likelihood of executing handler logic, but can cause a higher redis database
```
Comment (boosty) [inline lib/beetle/configuration.rb]: Still missing the "executing handler logic **more than once**" part. 

I would also add a short description why seeing a message more than twice could happen (e.g. due to a RabbitMQ redelivery or the publisher sending more than two copies).

File: test/beetle/message_test.rb (hunk 6 above)
Comment (skaes) [reply test/beetle/message_test.rb]: damn. fixed.

**Comment by skaes** [comment]
@boosty I think this is good to go into master. Agreed?

File: test/beetle/message_test.rb [hunk 8]
```diff
@@ -304,18 +304,17 @@ class SimpleMessageTest < Minitest::Test
     def setup
       @store = DeduplicationStore.new
       @store.flushdb
-      @store.expects(:redis).never
     end
 
-    test "when processing a simple message, ack should precede calling the handler" do
+    test "when processing a simple message, ack should folllow calling the handler" do
```
[the remark this answers is not in the document]
Comment (boosty) [reply test/beetle/message_test.rb]: Fixed it

**Pull request merged**